meterfs/meterfs.c

```diff
@@ -1311,3 +1312,146 @@ int meterfs_init(meterfs_ctx_t *ctx)
+/* Temporary adapters to implement storage fsops interfaces */
+
+
+static int adapter_open(void *info, oid_t *oid)
+{
+	int err;
+	struct _storage_devCtx_t *ctx = (struct _storage_devCtx_t *)info;
+
+	(void)mutexLock(ctx->lock);
+
+	err = meterfs_open(oid->id, &ctx->meterfsCtx);
+
+	(void)mutexUnlock(ctx->lock);
+	return err;
+}
+
+
+static int adapter_close(void *info, oid_t *oid)
+{
+	int err;
+	struct _storage_devCtx_t *ctx = (struct _storage_devCtx_t *)info;
+
+	(void)mutexLock(ctx->lock);
+
+	err = meterfs_close(oid->id, &ctx->meterfsCtx);
+
+	(void)mutexUnlock(ctx->lock);
+	return err;
+}
+
+
+static ssize_t adapter_read(void *info, oid_t *oid, offs_t offs, void *data, size_t len)
+{
+	int err;
+	struct _storage_devCtx_t *ctx = (struct _storage_devCtx_t *)info;
+
+	(void)mutexLock(ctx->lock);
+
+	err = meterfs_readFile(oid->id, offs, data, len, &ctx->meterfsCtx);
+
+	(void)mutexUnlock(ctx->lock);
+	return err;
+}
+
+
+static ssize_t adapter_write(void *info, oid_t *oid, offs_t offs, const void *data, size_t len)
+{
+	int err;
+	struct _storage_devCtx_t *ctx = (struct _storage_devCtx_t *)info;
+
+	(void)offs;
+
+	(void)mutexLock(ctx->lock);
+
+	err = meterfs_writeFile(oid->id, data, len, &ctx->meterfsCtx);
+
+	(void)mutexUnlock(ctx->lock);
+	return err;
+}
+
+
+static int adapter_devctl(void *info, oid_t *oid, const void *i, void *o)
+{
+	struct _storage_devCtx_t *ctx = (struct _storage_devCtx_t *)info;
+	const meterfs_i_devctl_t *in = (const meterfs_i_devctl_t *)i;
+	meterfs_o_devctl_t *out = (meterfs_o_devctl_t *)o;
+
+	(void)oid;
+
+	(void)mutexLock(ctx->lock);
+
+	out->err = meterfs_devctl(in, out, &ctx->meterfsCtx);
+
+	(void)mutexUnlock(ctx->lock);
+	return out->err;
+}
+
+
+static int adapter_lookup(void *info, oid_t *dir, const char *name, oid_t *res, oid_t *dev, char *lnk, int lnksz)
+{
+	int err;
+	struct _storage_devCtx_t *ctx = (struct _storage_devCtx_t *)info;
+
+	(void)lnk;
+	(void)lnksz;
+	(void)dev;
+
+	(void)mutexLock(ctx->lock);
+
+	res->port = dir->port;
+	err = meterfs_lookup(name, &res->id, &ctx->meterfsCtx);
+	*dev = *res;
+
+	(void)mutexUnlock(ctx->lock);
+	return err;
+}
+
+
+int meterfs_mount(storage_t *storage, storage_fs_t *fs, const char *data, unsigned long mode, oid_t *root)
+{
+	/* TODO implement missing functions. */
+	static const storage_fsops_t fsOps = {
+		.open = adapter_open,
+		.close = adapter_close,
+		.read = adapter_read,
+		.write = adapter_write,
+		.setattr = NULL,
+		.getattr = NULL,
+		.truncate = NULL,
+		.devctl = adapter_devctl,
+		.create = NULL,
+		.destroy = NULL,
+		.lookup = adapter_lookup,
+		.link = NULL,
+		.unlink = NULL,
+		.readdir = NULL,
+		.statfs = NULL,
+		.sync = NULL
+	};
+
+	(void)mode;
+	(void)data;
+
+	if ((storage == NULL) || (storage->dev == NULL) || (storage->dev->ctx == NULL) || (fs == NULL)) {
+		return -EINVAL;
+	}
+
+	root->id = 0;
+	fs->info = storage->dev->ctx;
+	fs->ops = &fsOps;
+
+	return EOK;
+}
+
+
+int meterfs_umount(storage_fs_t *fs)
+{
+	(void)fs;
```

Comment:
Missing resourceDestroy on mutex

Reply:
rn mutex is reused between consecutinves mounts/unmounts on the same storage